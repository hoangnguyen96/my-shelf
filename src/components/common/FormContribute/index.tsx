"use client";

import { useEffect, useMemo, useState } from "react";
import { Controller, useForm } from "react-hook-form";
import Image from "next/image";
import {
  Box,
  Flex,
  FormControl,
  FormErrorMessage,
  FormLabel,
  Select,
  Textarea,
} from "@chakra-ui/react";

// Constants
import { MESSAGES } from "@app/constants";

// Models
import { BookType } from "@app/models";

// Api
import { generateImageUpload } from "@app/api";

// Utils
import {
  clearErrorOnChange,
  formatDate,
  isEnableSubmitButton,
} from "@app/utils";
import { UploadIcon } from "@app/assets/icons";
import { Button, Input } from "..";

interface FormContributeProps {
  itemUpdate?: Partial<BookType>;
  onSubmit?: (book: Partial<BookType>) => void;
  onUpdate?: (id: string, book: Partial<BookType>) => void;
}

const FormContribute = ({
  itemUpdate,
  onSubmit,
  onUpdate,
}: FormContributeProps) => {
  const {
    id,
    title,
    author,
    category,
    imageUrl,
    description,
    createdDate,
    edition,
    publicationYear,
    rating,
    status,
  } = itemUpdate || {};
  const [selectedImage, setSelectedImage] = useState<File | null>(null);
  const [imagePreview, setImagePreview] = useState<string | null>(
    selectedImage ? URL.createObjectURL(selectedImage) : imageUrl || null
  );

  const REQUIRED_FIELDS = [
    "title",
    "author",
    "category",
    "imageUrl",
    "description",
  ];

  const formInitData: Partial<BookType> = useMemo(
    () => ({
      title: title || "",
      author: author || "",
      category: category || "",
      imageUrl: imageUrl || "",
      description: description || "",
    }),
    [title, author, category, description]
  );

  const {
    control,
    clearErrors,
    handleSubmit: submitForm,
    formState: { errors, isValid, dirtyFields },
    reset,
  } = useForm<Partial<BookType>>({
    mode: "onBlur",
    reValidateMode: "onBlur",
    defaultValues: formInitData,
  });

  const dirtyItems = Object.keys(dirtyFields);

  const shouldEnable = useMemo(
    () => isEnableSubmitButton(REQUIRED_FIELDS, dirtyItems, errors),
    [dirtyItems, errors]
  );

  const isDisableSubmit = !(shouldEnable || isValid);

  const handleFormSubmit = async (dataBook: Partial<BookType>) => {
    if (!selectedImage && !imageUrl) {
      console.error("No image selected");
      return;
    }

    try {
      if (itemUpdate) {
        let finalImageUrl = imageUrl;
        if (selectedImage) {
          const data = await generateImageUpload(selectedImage as File);
          if (data.success) {
            finalImageUrl = data.data.url;
          } else {
            console.error("Upload failed:", data.error.message);
            return;
          }
        }
        onUpdate?.(id as string, {
          ...dataBook,
          createdDate,
          edition,
          publicationYear,
          rating,
          status,
          imageUrl: finalImageUrl,
        });

        return;
      }

      const data = await generateImageUpload(selectedImage as File);

      if (data.success) {
        onSubmit?.({
          ...dataBook,
          status: false,
          publicationYear: new Date().getFullYear(),
          rating: 0,
          createdDate: formatDate(new Date()),
          edition: "first",
          imageUrl: data.data.url,
        });
        reset();
        setSelectedImage(null);
        setImagePreview(null);
        return;
      } else {
        console.error("Upload failed:", data.error.message);
        return;
      }
    } catch (error) {
      console.error("Error add book!", error);
      return;
    }
  };

  const handleFileChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const file = event.target.files?.[0];
    if (file) {
      setSelectedImage(file);
      setImagePreview(URL.createObjectURL(file));
    }
  };

  useEffect(() => {
    reset(formInitData);
  }, [itemUpdate, reset, formInitData]);

  return (
    <Box as="form" pos="relative">
      <Flex justifyContent="space-between" alignItems="center" gap="10px">
        {/* Book name */}
        <FormControl
          maxW={360}
          isInvalid={!!errors.title}
          mb={errors.title?.message ? "0" : "22px"}
        >
          <Controller
            name="title"
            control={control}
            defaultValue=""
            rules={{
              required: MESSAGES.FIELD_REQUIRED,
            }}
            render={({
              field: { value, onChange, ...rest },
              fieldState: { error },
            }) => (
              <Input
                id="title"
                value={value}
                placeholder="Book name"
                isInvalid={!!error?.message}
                onChange={(e) => {
                  onChange(e);
                  clearErrorOnChange("title", errors, clearErrors);
                }}
                {...rest}
              />
            )}
          />
          <FormErrorMessage pl="10px">
            {errors.title?.message && errors.title.message}
          </FormErrorMessage>
        </FormControl>

        {/* Category */}
        <FormControl
          maxW={170}
          w="100%"
          isInvalid={!!errors.category}
          mb={errors.category?.message ? "0" : "22px"}
        >
          <Controller
            name="category"
            control={control}
            defaultValue=""
            rules={{
              required: MESSAGES.FIELD_REQUIRED,
            }}
            render={({
              field: { value, onChange, ...rest },
              fieldState: { error },
            }) => (
              <Select
                placeholder="Category"
                border="2px"
                borderRadius="40px"
                value={value}
                w="100%"
                h="50px"
                borderColor="borderDefault"
                bgColor="white"
                _focusVisible={{ borderColor: "borderDefault" }}
                onChange={(e) => {
                  onChange(e);
                  clearErrorOnChange("category", errors, clearErrors);
                }}
              >
                <Box as="option" value="SelfHelpBook">
                  SelfHelpBook
                </Box>
                <Box as="option" value="TextBook">
                  TextBook
                </Box>
              </Select>
            )}
          />
          <FormErrorMessage pl="10px">
            {errors.category?.message && errors.category.message}
          </FormErrorMessage>
        </FormControl>
      </Flex>

      {/* Author name */}
      <FormControl
        maxW={360}
        isInvalid={!!errors.author}
        mb={errors.author?.message ? "0" : "22px"}
      >
        <Controller
          name="author"
          control={control}
          defaultValue=""
          rules={{
            required: MESSAGES.FIELD_REQUIRED,
          }}
          render={({
            field: { value, onChange, ...rest },
            fieldState: { error },
          }) => (
            <Input
              id="author"
              value={value}
              placeholder="Author Name"
              isInvalid={!!error?.message}
              onChange={(e) => {
                onChange(e);
                clearErrorOnChange("author", errors, clearErrors);
              }}
              {...rest}
            />
          )}
        />
        <FormErrorMessage pl="10px">
          {errors.author?.message && errors.author.message}
        </FormErrorMessage>
      </FormControl>

      {/* Upload Image */}
      <FormControl
        maxW={170}
        w="100%"
        isInvalid={!!errors.imageUrl}
        mb={errors.imageUrl?.message ? "0" : "12px"}
        pos="absolute"
        right={0}
        top={75}
      >
        <Controller
          name="imageUrl"
          control={control}
          defaultValue=""
          rules={{
            required: MESSAGES.FIELD_REQUIRED,
          }}
          render={({
            field: { value, onChange, ...rest },
            fieldState: { error },
          }) => (
            <>
              <FormLabel htmlFor="fileInput" cursor="pointer">
                {imagePreview ? (
                  <Image
                    src={imagePreview}
                    alt="Upload Image"
                    width={100}
                    height={100}
                    style={{ margin: "0 auto" }}
                  />
                ) : (
                  <Flex
                    justifyContent="center"
                    alignItems="center"
                    w={176}
                    h={63}
                    bgColor="backgroundUpload"
                    borderRadius="20px"
                  >
                    <UploadIcon />
                  </Flex>
                )}
              </FormLabel>
              <Input
                id="fileInput"
                type="file"
                accept="image/*"
                onChange={(e) => {
                  onChange(e);
                  handleFileChange(e);
                  clearErrorOnChange("author", errors, clearErrors);
                }}
                display="none"
              />
            </>
          )}
        />
        <FormErrorMessage pl="10px">
          {errors.imageUrl?.message && errors.imageUrl.message}
        </FormErrorMessage>
      </FormControl>

      {/* Reason */}
      <FormControl
        maxW={360}
        maxH={150}
        w="100%"
        h="100%"
        isInvalid={!!errors.description}
        mb={errors.description?.message ? "0" : "12px"}
      >
        <Controller
          name="description"
          control={control}
          defaultValue=""
          render={({
            field: { value, onChange, ...rest },
            fieldState: { error },
          }) => (
            <Textarea
              h="100%"
              value={value}
              border="1px solid"
              borderColor="borderDefault"
              placeholder="Reason For Your Contribution"
              _placeholder={{ color: "dark.60" }}
              _focusVisible={{ boxShadow: "unset" }}
              onChange={(e) => {
                onChange(e);
                clearErrorOnChange("author", errors, clearErrors);
              }}
              {...rest}
            />
          )}
        />
        <FormErrorMessage pl="10px">
          {errors.description?.message && errors.description.message}
        </FormErrorMessage>
      </FormControl>

      <Button
        type="submit"
        size="xl"
        text={itemUpdate ? "Update" : "Submit"}
        my="40px"
        isDisabled={isDisableSubmit}
        onClick={submitForm(handleFormSubmit)}
      />
    </Box>
  );
};

export default FormContribute;
