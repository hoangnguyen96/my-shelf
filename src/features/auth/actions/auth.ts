"use server";

import { AuthError } from "next-auth";
import { MESSAGES } from "@app/constants";
import { User } from "@app/interface";
import { signIn, signOut } from "@app/auth";

export const authenticate = async (formData: Partial<User>) => {
  try {
    await signIn("credentials", formData);
  } catch (error) {
    if (error instanceof AuthError) {
      return MESSAGES.EMAIL_PASSWORD_INVALID;
    }
    throw error;
  }
};

export const logout = async () => {
  await signOut();
};
