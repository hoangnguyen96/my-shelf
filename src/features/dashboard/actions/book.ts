import { BookType } from "@app/models";
import { API_ROUTES, MESSAGES } from "@app/constants";
import { api } from "@app/services";
import { dividePaginationBooks, getThreeTopBook } from "@app/utils";

export const getAllBook = async () => {
  try {
    const data = await api.get<BookType[]>(API_ROUTES.BOOKS);

    return data;
  } catch (error) {
    if (error instanceof Error) {
      return error.message;
    }
    return MESSAGES.RESPONSE_ERROR;
  }
};

export const getTopThreeBook = async () => {
  try {
    const data = await api.get<BookType[]>(API_ROUTES.BOOKS, {
      next: { revalidate: 60 },
    });
    const result = getThreeTopBook(data);

    return result;
  } catch (error) {
    if (error instanceof Error) {
      return error.message;
    }
    return MESSAGES.RESPONSE_ERROR;
  }
};

export const getPaginatedBook = async (params?: string) => {
  try {
    const data = await api.get<BookType[]>(
      `${API_ROUTES.BOOKS}?${params || ""}`,
      {
        cache: "no-store",
      }
    );
    const result = dividePaginationBooks(data);

    return result;
  } catch (error) {
    if (error instanceof Error) {
      return error.message;
    }
    return MESSAGES.RESPONSE_ERROR;
  }
};

export const getTwelveItemBook = async (params?: string) => {
  try {
    const data = await api.get<BookType[]>(
      `${API_ROUTES.BOOKS}?${params || ""}page=1&limit=12`,
      { next: { revalidate: 3600 } }
    );

    return data;
  } catch (error) {
    if (error instanceof Error) {
      return error.message;
    }
    return MESSAGES.RESPONSE_ERROR;
  }
};

export const getBookByParams = async (params: string) => {
  try {
    const data = await api.get<BookType[]>(`${API_ROUTES.BOOKS}?${params}`);

    return data;
  } catch (error) {
    if (error instanceof Error) {
      return error.message;
    }
    return MESSAGES.RESPONSE_ERROR;
  }
};

export const getBookById = async (id: string) => {
  try {
    const data = await api.get<BookType>(`${API_ROUTES.BOOKS}/${parseInt(id)}`);

    return data;
  } catch (error) {
    if (error instanceof Error) {
      return error.message;
    }
    return MESSAGES.RESPONSE_ERROR;
  }
};

export const addBook = async (payload: Partial<BookType>) => {
  return await api.post(API_ROUTES.BOOKS, payload);
};

export const updateBookById = async (
  id: string,
  payload: Partial<BookType>
) => {
  return await api.put(`${API_ROUTES.BOOKS}/${id}`, payload);
};

export const deleteBook = async (id: string) => {
  return await api.delete(`${API_ROUTES.BOOKS}/${id}`);
};
