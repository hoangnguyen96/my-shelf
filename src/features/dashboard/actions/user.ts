import { API_ROUTES, MESSAGES } from "@app/constants";
import { User } from "@app/models";
import { api } from "@app/services";

export const getAllUser = async () => {
  try {
    const data = await api.get<User[]>(API_ROUTES.USER);

    return data || [];
  } catch (error) {
    if (error instanceof Error) {
      return error.message;
    }
    return MESSAGES.RESPONSE_ERROR;
  }
};

export const getUserById = async (id: string) => {
  try {
    const data = await api.get<User[]>(`${API_ROUTES.USER}?userId=${id}`);

    return data[0];
  } catch (error) {
    if (error instanceof Error) {
      return error.message;
    }
    return MESSAGES.RESPONSE_ERROR;
  }
};

export const getUserByEmail = async (email: string) => {
  try {
    const user = (await api.get(`${API_ROUTES.USER}?email=${email}`)) as User[];

    return user;
  } catch (error) {
    if (error instanceof Error) {
      return error.message;
    }
    return MESSAGES.RESPONSE_ERROR;
  }
};

export const updateUserById = async (id: string, payload: Partial<User>) => {
  return await api.put(`${API_ROUTES.USER}/${id}`, payload);
};

export const addUser = async (payload: Partial<User>) => {
  return await api.post(API_ROUTES.USER, payload);
};
