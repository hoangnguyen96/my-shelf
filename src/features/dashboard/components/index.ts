export * from "./home-list";
export * from "./search-list";
export * from "./my-book-shelf";
export * from "./my-book-shelf-favorites";
export * from "./preview-book-details";
export * from "./profile-update";
export * from "./contribute";
export * from "./contribute-update";
export * from "./contribute-complete";
export * from "./contribute-list";
