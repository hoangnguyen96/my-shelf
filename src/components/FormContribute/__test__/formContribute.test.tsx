import "@testing-library/jest-dom";
import { act, fireEvent, render, waitFor } from "@testing-library/react";
import FormContribute from "..";
import { generateImageUpload } from "@app/features/dashboard/actions";

// Mock Image component
jest.mock("next/image", () => ({
  __esModule: true,
  default: (props: any) => <img {...props} />,
}));

jest.mock("@app/features/dashboard/actions", () => ({
  generateImageUpload: jest.fn().mockResolvedValue({
    success: true,
    url: "https://example.com/image.jpg",
  }),
}));

describe("Form Contribute", () => {
  const mockOnUpdate = jest.fn();
  const mockOnSubmit = jest.fn();
  const props = {
    itemUpdate: {
      title: "Don't Make Me Think",
      author: "Steve Krug",
      category: "SelfHelpBook",
      imageUrl: "https://i.ibb.co/QbQgVtG/book1.png",
      description: "description will update later",
      status: false,
      publicationYear: 2000,
      rating: 4.5,
      createdDate: "3 Sep 2024 9:07 PM",
      edition: "first",
      id: "1",
    },
    onSubmit: mockOnSubmit,
    onUpdate: mockOnUpdate,
  };

  it("Should render correctly snapshot", () => {
    expect(render(<FormContribute {...props} />)).toMatchSnapshot();
  });

  it("should render the form correctly with initial data", () => {
    const { getByPlaceholderText, getByTestId } = render(
      <FormContribute {...props} />
    );

    expect(getByPlaceholderText("Book name")).toHaveValue(
      "Don't Make Me Think"
    );
    expect(getByPlaceholderText("Author Name")).toHaveValue("Steve Krug");
    expect(getByPlaceholderText("Reason For Your Contribution")).toHaveValue(
      "description will update later"
    );
    expect(getByTestId("contribute-category")).toHaveValue("SelfHelpBook");
  });

  it("should update form fields on input change", () => {
    const { getByPlaceholderText, getByTestId } = render(
      <FormContribute {...props} />
    );

    const titleInput = getByPlaceholderText("Book name");
    fireEvent.change(titleInput, { target: { value: "Updated Book Title" } });
    expect(titleInput).toHaveValue("Updated Book Title");

    const authorInput = getByPlaceholderText("Author Name");
    fireEvent.change(authorInput, { target: { value: "Updated Author" } });
    expect(authorInput).toHaveValue("Updated Author");
  });

  it("should handle file input changes", async () => {
    (generateImageUpload as jest.Mock).mockResolvedValueOnce({
      success: true,
      data: {
        url: "http://mocked-image-url.com",
      },
    });

    const { getByTestId, getByAltText } = render(<FormContribute {...props} />);

    const mockURL = "https://i.ibb.co/QbQgVtG/book1.png";
    URL.createObjectURL = jest.fn(() => mockURL);

    const file = new File(["image"], "image.png", { type: "image/png" });

    // Trigger file input change event
    const input = getByTestId("input-file-contribute");
    fireEvent.change(input, { target: { files: [file] } });

    await waitFor(() => {
      expect(getByAltText("Upload Image")).toHaveAttribute("src", mockURL);
    });
  });

  // TODO: Will update later
  it("Should handle update book successful", async () => {
    const { getByPlaceholderText, getByTestId, getByText } = render(
      <FormContribute {...props} />
    );

    await act(() => {
      fireEvent.change(getByPlaceholderText("Book name"), {
        target: { value: "Test Book" },
      });
      fireEvent.change(getByPlaceholderText("Author Name"), {
        target: { value: "Test Author" },
      });
      fireEvent.change(getByPlaceholderText("Reason For Your Contribution"), {
        target: { value: "Test Description" },
      });
      const categorySelect = getByTestId("contribute-category");

      fireEvent.change(categorySelect, { target: { value: "SelfHelpBook" } });

      // const mockURL = "https://i.ibb.co/QbQgVtG/book1.png";
      // URL.createObjectURL = jest.fn(() => mockURL);
      // const input = getByTestId("input-file-contribute");
      // fireEvent.change(input, {
      //   target: {
      //     files: [new File(["image"], "image.png", { type: "image/png" })],
      //   },
      // });
      // (generateImageUpload as jest.Mock).mockResolvedValueOnce({
      //   success: true,
      //   data: {
      //     url: "http://mocked-image-url.com",
      //   },
      // });

      // // const { getByTestId, getByAltText } = render(
      // //   <FormContribute {...props} />
      // // );

      // const mockURL = "https://i.ibb.co/QbQgVtG/book5.png";
      // URL.createObjectURL = jest.fn(() => mockURL);

      // const file = new File(["image"], "image.png", { type: "image/png" });

      // // Trigger file input change event
      // const input = getByTestId("input-file-contribute");
      // fireEvent.change(input, { target: { files: [file] } });
    });

    fireEvent.click(getByTestId("submit-contribute"));

    // await waitFor(() => {
    //   expect(mockOnSubmit).toHaveBeenCalledWith({
    //     title: "Test Book",
    //     author: "Test Author",
    //     category: "SelfHelpBook",
    //     imageUrl: "https://i.ibb.co/QbQgVtG/book1.png",
    //     description: "Test Description",
    //   });
    // });
  });

  it("Should handle add book successful", async () => {
    const { getByPlaceholderText, getByTestId, getByRole } = render(
      <FormContribute />
    );

    await act(async () => {
      fireEvent.change(getByPlaceholderText("Book name"), {
        target: { value: "Test Book" },
      });
      fireEvent.change(getByPlaceholderText("Author Name"), {
        target: { value: "Test Author" },
      });
      fireEvent.change(getByPlaceholderText("Reason For Your Contribution"), {
        target: { value: "Test Description" },
      });
      const categorySelect = getByTestId("contribute-category");

      fireEvent.change(categorySelect, { target: { value: "SelfHelpBook" } });
      (generateImageUpload as jest.Mock).mockResolvedValueOnce({
        success: true,
        data: {
          url: "http://mocked-image-url.com",
        },
      });
      // const mockURL = "https://i.ibb.co/QbQgVtG/book1.png";
      // URL.createObjectURL = jest.fn(() => mockURL);
      // const input = getByTestId("input-file-contribute");
      // fireEvent.change(input, {
      //   target: {
      //     files: [new File(["image"], "image.png", { type: "image/png" })],
      //   },
      // });
    });

    fireEvent.click(getByTestId("submit-contribute"));
    // await act(async () => {
    // });
  });

  it("Should handle add book failed", async () => {
    const { getByPlaceholderText, getByTestId, getByRole } = render(
      <FormContribute />
    );

    await act(async () => {
      fireEvent.change(getByPlaceholderText("Book name"), {
        target: { value: "Test Book" },
      });
      fireEvent.change(getByPlaceholderText("Author Name"), {
        target: { value: "Test Author" },
      });
      fireEvent.change(getByPlaceholderText("Reason For Your Contribution"), {
        target: { value: "Test Description" },
      });
      const categorySelect = getByTestId("contribute-category");

      fireEvent.change(categorySelect, { target: { value: "SelfHelpBook" } });
      (generateImageUpload as jest.Mock).mockResolvedValueOnce({
        success: true,
        data: {
          url: "http://mocked-image-url.com",
        },
      });
      // const mockURL = "https://i.ibb.co/QbQgVtG/book1.png";
      // URL.createObjectURL = jest.fn(() => mockURL);
      // const input = getByTestId("input-file-contribute");
      // fireEvent.change(input, {
      //   target: {
      //     files: [new File(["image"], "image.png", { type: "image/png" })],
      //   },
      // });
    });

    fireEvent.click(getByTestId("submit-contribute"));
    // await act(async () => {
    // });
  });
});
