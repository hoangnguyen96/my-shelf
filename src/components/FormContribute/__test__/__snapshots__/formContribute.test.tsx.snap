// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Should render correctly snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="css-79elbk"
      >
        <div
          class="css-92ksox"
        >
          <div
            class="chakra-form-control css-2d82ls"
            role="group"
          >
            <div
              class="chakra-input__group css-bx0blc"
              data-group="true"
            >
              <input
                class="chakra-input css-1x36n8t"
                id="title"
                name="title"
                placeholder="Book name"
                type="text"
                value="Don't Make Me Think"
              />
            </div>
          </div>
          <div
            class="chakra-form-control css-106bepj"
            role="group"
          >
            <div
              class="chakra-select__wrapper css-42b2qy"
            >
              <select
                class="chakra-select css-g3rz38"
                data-testid="contribute-category"
                id="field-:r1:"
              >
                <option
                  value=""
                >
                  Category
                </option>
                <option
                  class="css-0"
                  value="SelfHelpBook"
                >
                  SelfHelpBook
                </option>
                <option
                  class="css-0"
                  value="TextBook"
                >
                  TextBook
                </option>
              </select>
              <div
                class="chakra-select__icon-wrapper css-162mkon"
              >
                <svg
                  aria-hidden="true"
                  class="chakra-select__icon"
                  focusable="false"
                  role="presentation"
                  style="width: 1em; height: 1em; color: currentColor;"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                    fill="currentColor"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
        <div
          class="chakra-form-control css-2d82ls"
          role="group"
        >
          <div
            class="chakra-input__group css-bx0blc"
            data-group="true"
          >
            <input
              class="chakra-input css-1x36n8t"
              id="author"
              name="author"
              placeholder="Author Name"
              type="text"
              value="Steve Krug"
            />
          </div>
        </div>
        <div
          class="chakra-form-control css-uzlds2"
          role="group"
        >
          <label
            class="chakra-form__label css-zfpjyo"
            data-testid="fileInput"
            for="fileInput"
            id="field-:r3:-label"
          >
            <img
              alt="Upload Image"
              height="100"
              src="https://i.ibb.co/QbQgVtG/book1.png"
              style="margin: 0px auto;"
              width="100"
            />
          </label>
          <div
            class="chakra-input__group css-bx0blc"
            data-group="true"
          >
            <input
              accept="image/*"
              class="chakra-input css-1vuqmsh"
              data-testid="input-file-contribute"
              id="fileInput"
              type="file"
            />
          </div>
        </div>
        <div
          class="chakra-form-control css-8qt7d5"
          role="group"
        >
          <textarea
            class="chakra-textarea css-1yn82jt"
            id="field-:r4:"
            name="description"
            placeholder="Reason For Your Contribution"
          >
            description will update later
          </textarea>
        </div>
        <button
          class="chakra-button css-1qhatj0"
          data-testid="submit-contribute"
          disabled=""
          type="submit"
        >
          Update
        </button>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="css-79elbk"
    >
      <div
        class="css-92ksox"
      >
        <div
          class="chakra-form-control css-2d82ls"
          role="group"
        >
          <div
            class="chakra-input__group css-bx0blc"
            data-group="true"
          >
            <input
              class="chakra-input css-1x36n8t"
              id="title"
              name="title"
              placeholder="Book name"
              type="text"
              value="Don't Make Me Think"
            />
          </div>
        </div>
        <div
          class="chakra-form-control css-106bepj"
          role="group"
        >
          <div
            class="chakra-select__wrapper css-42b2qy"
          >
            <select
              class="chakra-select css-g3rz38"
              data-testid="contribute-category"
              id="field-:r1:"
            >
              <option
                value=""
              >
                Category
              </option>
              <option
                class="css-0"
                value="SelfHelpBook"
              >
                SelfHelpBook
              </option>
              <option
                class="css-0"
                value="TextBook"
              >
                TextBook
              </option>
            </select>
            <div
              class="chakra-select__icon-wrapper css-162mkon"
            >
              <svg
                aria-hidden="true"
                class="chakra-select__icon"
                focusable="false"
                role="presentation"
                style="width: 1em; height: 1em; color: currentColor;"
                viewBox="0 0 24 24"
              >
                <path
                  d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                  fill="currentColor"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
      <div
        class="chakra-form-control css-2d82ls"
        role="group"
      >
        <div
          class="chakra-input__group css-bx0blc"
          data-group="true"
        >
          <input
            class="chakra-input css-1x36n8t"
            id="author"
            name="author"
            placeholder="Author Name"
            type="text"
            value="Steve Krug"
          />
        </div>
      </div>
      <div
        class="chakra-form-control css-uzlds2"
        role="group"
      >
        <label
          class="chakra-form__label css-zfpjyo"
          data-testid="fileInput"
          for="fileInput"
          id="field-:r3:-label"
        >
          <img
            alt="Upload Image"
            height="100"
            src="https://i.ibb.co/QbQgVtG/book1.png"
            style="margin: 0px auto;"
            width="100"
          />
        </label>
        <div
          class="chakra-input__group css-bx0blc"
          data-group="true"
        >
          <input
            accept="image/*"
            class="chakra-input css-1vuqmsh"
            data-testid="input-file-contribute"
            id="fileInput"
            type="file"
          />
        </div>
      </div>
      <div
        class="chakra-form-control css-8qt7d5"
        role="group"
      >
        <textarea
          class="chakra-textarea css-1yn82jt"
          id="field-:r4:"
          name="description"
          placeholder="Reason For Your Contribution"
        >
          description will update later
        </textarea>
      </div>
      <button
        class="chakra-button css-1qhatj0"
        data-testid="submit-contribute"
        disabled=""
        type="submit"
      >
        Update
      </button>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
